src/effects/VST3/VST3Utils.cpp

```diff
@@ -238,3 +245,132 @@ bool VST3Utils::ParseAutomationParameterKey(const wxString& key, Steinberg::Vst:
    return false;
 
 }
+
+
+const std::vector<VST3Wrapper::ParameterInfo>* VST3Wrapper::getParameterInfos() const
+{
+   if (mParameterInfos.size() == 0)
+   {
+      if (mEditController == nullptr)
+         return nullptr;
+
+      for (int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+      {
+         Steinberg::Vst::ParameterInfo vstParameterInfo{ };
+         if (mEditController->getParameterInfo(i, vstParameterInfo) == Steinberg::kResultOk)
+         {
+            // Add all parameterinfos, regardless of whether they can be automated or not,
+            // because different flags of them will be checked by different methods.
+            mParameterInfos.push_back(vstParameterInfo);
+         }
+      }
+   }
+
+   return &mParameterInfos;
+}
+
+
+bool VST3Wrapper::ForEachParameter(ParameterVisitor visitor) const
+{
+   const auto pParameterInfos = getParameterInfos();
+   if (pParameterInfos == nullptr)
+      return false;
+ 
+   for (const auto& parameterInfo : *pParameterInfos)
+   {
+      if ( ! visitor(parameterInfo) )
+         break;
+   }
+
+   return true;
+}
+
+
+bool VST3Wrapper::AtLeastOne(ParameterVisitor visitor) const
+{
+   const auto pParameterInfos = getParameterInfos();
+   if (pParameterInfos == nullptr)
+      return false;
+
+   for (const auto& parameterInfo : *pParameterInfos)
+   {
+      if (visitor(parameterInfo))
+      {
+         return true;
+      }
+   }
+
+   return false;
+}
+
+
+bool VST3Wrapper::LoadPreset(Steinberg::IBStream* fileStream, const Steinberg::FUID& classID)
+{
+   using namespace Steinberg;
+
+   return Vst::PresetFile::loadPreset
+   (
+      fileStream,
+      FUID::fromTUID(classID),
+      mEffectComponent.get(),
+      mEditController.get()
+   );
+}
+
+
+bool VST3Wrapper::SavePreset(Steinberg::IBStream* fileStream, const Steinberg::FUID& classID) const
+{
+   using namespace Steinberg;
+
+   return Vst::PresetFile::savePreset
+   (
+      fileStream,
+      classID,
+      mEffectComponent.get(),
+      mEditController.get()
+   );
+}
+
+
+bool VST3Wrapper::FetchSettings(VST3EffectSettings& settings) const
+{
+   settings.mValues.clear();
+
+   return ForEachParameter(
+
+      [&](const ParameterInfo& pi)
+      {
+         const auto id = pi.id;
+         settings.mValues[id] = mEditController->getParamNormalized(id);
+         return true;
+      }
+   );
+}
+
+
+bool VST3Wrapper::StoreSettings(const VST3EffectSettings& settings) const
+{
+   if (mEditController == nullptr)
+      return false;
+
+   return ForEachParameter
+   (
+      [&](const ParameterInfo& parameterinfo)
+      {
+         auto itr = settings.mValues.find(parameterinfo.id);
+         if (itr != settings.mValues.end())
+         {
+            const auto& value = itr->second;
+
+            mEditController->setParamNormalized(parameterinfo.id, value);
+         }
+         else
+         {
+            return false;
```

Comment:
You cause failure if a parameter is known in the plug-in but not known in the mapping.  This is not like what I did for AudioUnits.  I think it's better to ignore this case and continue storing other parameters.

Reply:
Alright, also done now